rust/lance-table/src/rowids/encoded_array.rs

```diff
@@ -0,0 +1,398 @@
+    /// u64 values represented as u16 offset from a base value.
+    ///
+    /// Useful when the min and max value are within u16 range (0..65535).
+    /// Only space saving when there are more than 2 values.
+    U16 { base: u64, offsets: Vec<u16> },
+    /// u64 values represented as u32 offset from a base value.
+    ///
+    /// Useful when the min and max value are within u32 range (0..~4 billion).
+    U32 { base: u64, offsets: Vec<u32> },
+    /// Just a plain vector of u64 values.
+    ///
+    /// For when the values cover a wide range.
+    U64(Vec<u64>),
+}
+
+impl EncodedU64Array {
+    pub fn len(&self) -> usize {
+        match self {
+            Self::U16 { offsets, .. } => offsets.len(),
+            Self::U32 { offsets, .. } => offsets.len(),
+            Self::U64(values) => values.len(),
+        }
+    }
+
+    pub fn iter(&self) -> Box<dyn DoubleEndedIterator<Item = u64> + '_> {
+        match self {
+            Self::U16 { base, offsets } => {
+                Box::new(offsets.iter().cloned().map(move |o| base + o as u64))
+            }
+            Self::U32 { base, offsets } => {
+                Box::new(offsets.iter().cloned().map(move |o| base + o as u64))
+            }
+            Self::U64(values) => Box::new(values.iter().cloned()),
+        }
+    }
+
+    pub fn get(&self, i: usize) -> Option<u64> {
+        match self {
+            Self::U16 { base, offsets } => {
+                if i < offsets.len() {
+                    Some(*base + offsets[i] as u64)
+                } else {
+                    None
+                }
+            }
+            Self::U32 { base, offsets } => {
+                if i < offsets.len() {
+                    Some(*base + offsets[i] as u64)
+                } else {
+                    None
+                }
+            }
+            Self::U64(values) => values.get(i).copied(),
+        }
+    }
+
+    pub fn min(&self) -> Option<u64> {
+        match self {
+            Self::U16 { base, offsets } => {
+                if offsets.is_empty() {
+                    None
+                } else {
+                    Some(*base)
+                }
+            }
+            Self::U32 { base, offsets } => {
+                if offsets.is_empty() {
+                    None
+                } else {
+                    Some(*base)
+                }
+            }
+            Self::U64(values) => values.iter().copied().min(),
+        }
+    }
+
+    pub fn max(&self) -> Option<u64> {
+        match self {
+            Self::U16 { base, offsets } => {
+                if offsets.is_empty() {
+                    None
+                } else {
+                    Some(*base + *offsets.last().unwrap() as u64)
+                }
+            }
+            Self::U32 { base, offsets } => {
+                if offsets.is_empty() {
+                    None
+                } else {
+                    Some(*base + *offsets.last().unwrap() as u64)
```

Comment:
is `last` guaranteed to be the max?

Reply:
Nope :) I'll change the unit test to not use sorted data.